File: cat-core/src/main/java/com/dianping/cat/message/internal/DefaultMessageProducer.java
package com.dianping.cat.message.internal;

import java.io.PrintWriter;
import java.io.StringWriter;

import org.unidal.lookup.annotation.Inject;

import com.dianping.cat.message.Event;
import com.dianping.cat.message.Heartbeat;
import com.dianping.cat.message.Message;
import com.dianping.cat.message.MessageProducer;
import com.dianping.cat.message.Metric;
import com.dianping.cat.message.Transaction;
import com.dianping.cat.message.spi.MessageManager;

public class DefaultMessageProducer implements MessageProducer {
	@Inject
	private MessageManager m_manager;

	@Inject
	private MessageIdFactory m_factory;

	@Override
	public String createMessageId() {
		return m_factory.getNextId();
	}

	@Override
	public boolean isEnabled() {
		return m_manager.isCatEnabled();
	}

	@Override
	public void logError(Throwable cause) {
		StringWriter writer = new StringWriter(2048);

		cause.printStackTrace(new PrintWriter(writer));

		if (cause instanceof Error) {
			logEvent("Error", cause.getClass().getName(), "ERROR", writer.toString());
		} else if (cause instanceof RuntimeException) {
			logEvent("RuntimeException", cause.getClass().getName(), "ERROR", writer.toString());
		} else {
			logEvent("Exception", cause.getClass().getName(), "ERROR", writer.toString());
		}
	}

	@Override
	public void logEvent(String type, String name) {
		logEvent(type, name, Message.SUCCESS, null);
	}

	@Override
	public void logEvent(String type, String name, String status, String nameValuePairs) {
		Event event = newEvent(type, name);

		if (nameValuePairs != null && nameValuePairs.length() > 0) {
			event.addData(nameValuePairs);
		}

		event.setStatus(status);
		event.complete();
	}

	@Override
	public void logHeartbeat(String type, String name, String status, String nameValuePairs) {
		Heartbeat heartbeat = newHeartbeat(type, name);

		heartbeat.addData(nameValuePairs);
		heartbeat.setStatus(status);
		heartbeat.complete();
	}

	@Override
   public void logMetric(String type, String name, String status, String nameValuePairs) {
		Metric event = newMetric(type, name);

		if (nameValuePairs != null && nameValuePairs.length() > 0) {
			event.addData(nameValuePairs);
		}

		event.setStatus(status);
		event.complete();
   }

	@Override
	public Event newEvent(String type, String name) {
		if (!m_manager.hasContext()) {
			m_manager.setup();
		}

		if (m_manager.isCatEnabled()) {
			DefaultEvent event = new DefaultEvent(type, name);

			m_manager.add(event);
			return event;
		} else {
			return NullMessage.EVENT;
		}
	}
	
	public Event newEvent(Transaction parent, String type, String name) {
		if (!m_manager.hasContext()) {
			m_manager.setup();
		}

		if (m_manager.isCatEnabled() && parent != null) {
			DefaultEvent event = new DefaultEvent(type, name);

			parent.addChild(event);
			return event;
		} else {
			return NullMessage.EVENT;
		}
	}

	@Override
	public Heartbeat newHeartbeat(String type, String name) {
		if (!m_manager.hasContext()) {
			m_manager.setup();
		}

		if (m_manager.isCatEnabled()) {
			DefaultHeartbeat heartbeat = new DefaultHeartbeat(type, name);

			m_manager.add(heartbeat);
			return heartbeat;
		} else {
			return NullMessage.HEARTBEAT;
		}
	}

	public Heartbeat newHeartbeat(Transaction parent, String type, String name) {
		if (!m_manager.hasContext()) {
			m_manager.setup();
		}

		if (m_manager.isCatEnabled() && parent != null) {
			DefaultHeartbeat heartbeat = new DefaultHeartbeat(type, name);

			parent.addChild(heartbeat);
			return heartbeat;
		} else {
			return NullMessage.HEARTBEAT;
		}
	}

	@Override
	public Metric newMetric(String type, String name) {
		if (!m_manager.hasContext()) {
			m_manager.setup();
		}
		
		if (m_manager.isCatEnabled()) {
			DefaultMetric metric = new DefaultMetric(type, name);
			
			m_manager.add(metric);
			return metric;
		} else {
			return NullMessage.METRIC;
		}
	}

	@Override
	public Transaction newTransaction(String type, String name) {
		// this enable CAT client logging cat message without explicit setup
		if (!m_manager.hasContext()) {
			m_manager.setup();
		}

		if (m_manager.isCatEnabled()) {
			DefaultTransaction transaction = new DefaultTransaction(type, name, m_manager);

			m_manager.start(transaction);
			return transaction;
		} else {
			return NullMessage.TRANSACTION;
		}
	}

	public Transaction newTransaction(Transaction parent, String type, String name) {
		// this enable CAT client logging cat message without explicit setup
		if (!m_manager.hasContext()) {
			m_manager.setup();
		}

		if (m_manager.isCatEnabled() && parent != null) {
			DefaultTransaction transaction = new DefaultTransaction(type, name, m_manager);

			parent.addChild(transaction);
			transaction.setStandalone(false);
			return transaction;
		} else {
			return NullMessage.TRANSACTION;
		}
	}
}

File: lib/csharp/src/CatClient/Message/Spi/Internals/DefaultMessageProducer.cs
using Org.Unidal.Cat.Message.Internals;
using System;
using System.IO;
using System.Collections.Generic;
using System.Text;
using Org.Unidal.Cat.Util;

namespace Org.Unidal.Cat.Message.Spi.Internals
{
    public class DefaultMessageProducer : IMessageProducer
    {
        private const string TAG_MAPS_SEPARATOR = "@|";
        private readonly IMessageManager _mManager;

        public DefaultMessageProducer(IMessageManager manager)
        {
            _mManager = manager;
        }

        #region IMessageProducer Members

        public virtual void LogError(Exception cause)
        {
            LogError(null, cause);
        }

        public virtual void LogError(String message, Exception cause)
        {
            if (!ShouldLog(cause))
            {
                 //Console.WriteLine("Ignore error log of cause: " + cause);
                return;
            }

            StringWriter writer = new StringWriter();
            if (!String.IsNullOrWhiteSpace(message))
            {
                writer.WriteLine(message);
            }
            writer.WriteLine(cause.ToString());
            var exceptionType = (cause.Source !=null && cause.Source.Length > 0 && cause.Source[0] == '~') ?
                cause.Source.Substring(1) : cause.GetType().FullName;
            LogEvent("Error", exceptionType, "ERROR",
                    writer.ToString());
        }

        public virtual void LogEvent(String type, String name, String status, String nameValuePairs)
        {
            IEvent evt0 = NewEvent(type, name, nameValuePairs);

            evt0.Status = status;
            evt0.Complete();
        }

        public virtual void LogHeartbeat(String type, String name, String status, String nameValuePairs)
        {
            IHeartbeat heartbeat = NewHeartbeat(type, name, nameValuePairs);
            heartbeat.Status = status;
            heartbeat.Complete();
        }

        public virtual void LogMetric(String name, String status, String nameValuePairs)
        {
            string type = "";
            IMetric metric = NewMetric(type, name, nameValuePairs);
            metric.Status = status;
            metric.Complete();
        }

        public virtual IEvent NewEvent(String type, String name, String nameValuePairs)
        {
            if (!_mManager.HasContext())
            {
                _mManager.Setup();
            }

            if (_mManager.CatEnabled)
            {
                IEvent evt0 = new DefaultEvent(type, name);

                // Need add data before event is added into context, so that estimated bytes is correct.
                if (!string.IsNullOrEmpty(nameValuePairs))
                {
                    evt0.AddData(nameValuePairs);
                }

                _mManager.Add(evt0);
                return evt0;
            }
            return new NullEvent();
        }

        public virtual IMetric NewMetric(String type, String name, String nameValuePairs)
        {
            if (!_mManager.HasContext())
            {
                _mManager.Setup();
            }

            if (_mManager.CatEnabled)
            {
                IMetric metric = new DefaultMetric(type, name);
                if (!string.IsNullOrEmpty(nameValuePairs))
                {
                    metric.AddData(nameValuePairs);
                }
                _mManager.Add(metric);
                return metric;
            }
            return new NullMetric();
        }

        public virtual IHeartbeat NewHeartbeat(String type, String name, String nameValuePairs)
        {
            if (!_mManager.HasContext())
            {
                _mManager.Setup();
            }

            if (_mManager.CatEnabled)
            {
                IHeartbeat heartbeat = new DefaultHeartbeat(type, name);
                if (!string.IsNullOrEmpty(nameValuePairs))
                {
                    heartbeat.AddData(nameValuePairs);
                }
                _mManager.Add(heartbeat);
                return heartbeat;
            }
            return new NullHeartbeat();
        }

        public virtual ITransaction NewTransaction(String type, String name)
        {
            // this enable CAT client logging cat message without explicit setup
            if (!_mManager.HasContext())
            {
                _mManager.Setup();
            }

            if (_mManager.CatEnabled)
            {
                ITransaction transaction = new DefaultTransaction(type, name, _mManager);
                _mManager.Start(transaction, false);
                return transaction;
            }
            return new NullTransaction();
        }

        public virtual IForkedTransaction NewForkedTransaction(String type, String name)
        {
            if (!_mManager.HasContext())
            {
                _mManager.Setup();
            }

            if (_mManager.CatEnabled)
            {
                DefaultForkedTransaction transaction = new DefaultForkedTransaction(type, name, _mManager);
                if (_mManager is DefaultMessageManager)
                {
                    ((DefaultMessageManager)_mManager).LinkAsRunAway(transaction);
                }
                _mManager.Start(transaction, true);
                return transaction;
            }
            else
            {
                return new NullTransaction();
            }

        }

        public virtual ITaggedTransaction NewTaggedTransaction(string type, string name, string tag)
        {
            if (!_mManager.HasContext())
            {
                _mManager.Setup();
            }

            if (_mManager.CatEnabled)
            {
                DefaultTaggedTransaction transaction = new DefaultTaggedTransaction(type, name, tag, _mManager);
                _mManager.Start(transaction, true);
                return transaction;
            }
            else
            {
                return new NullTransaction();
            }
        }

        private bool ShouldLog(Exception e)
        {
            if (_mManager is DefaultMessageManager)
            {
                return ((DefaultMessageManager)_mManager).ShouldLog(e);
            }
            else
            {
                return true;
            }
        }

        #endregion
    }
}